### PR TITLE
Poisson node

### PR DESCRIPTION
**Pull request by semihakbayrak**
For calculating the average energy and differential entropy, we referred the following sources: https://arxiv.org/pdf/1708.06394.pdf
https://en.wikipedia.org/wiki/Poisson_distribution

### CODE REVIEW

File: src/factor_nodes/poisson.jl [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+ForneyLab.vague(::Type{Poisson}) = ProbabilityDistribution(Univariate, Poisson, l=huge)
+
+ForneyLab.isProper(dist::ProbabilityDistribution{Univariate, Poisson}) = (0 < dist.params[:l] < huge)
+
+ForneyLab.unsafeMean(dist::ProbabilityDistribution{Univariate, Poisson}) = dist.params[:l]
+
+ForneyLab.unsafeVar(dist::ProbabilityDistribution{Univariate, Poisson}) = dist.params[:l]
+
+ForneyLab.unsafeLogMean(dist::ProbabilityDistribution{Univariate, Poisson}) = dist.params[:l]
```
Comment (ThijsvdLaar) [inline src/factor_nodes/poisson.jl]: Please verify this expression, is there a reference for this result? Also, this expression does not appear in the tests yet; please add a test.

File: src/factor_nodes/poisson.jl [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    l = clamp(dist.params[:l], tiny, huge)
+    h(l, j=20) = sum([(l)^(k)*log(factorial(k))/factorial(k) for k in collect(0:j)])
+    l*(1-log(l)) + exp(-l)*h(l)
+end
+
+# Average energy functional
+function ForneyLab.averageEnergy(::Type{Poisson}, marg_out::ProbabilityDistribution{Univariate}, marg_l::ProbabilityDistribution{Univariate})
+    # approximation of expectation of logX!
+    h(j) = sum([binomial(j, k)*(-1)^(-k)*log(factorial(k)) for k in collect(0:j)])
+    Elog_fact(λ, lim=20) = sum([(-λ)^j/factorial(j)*h(j) for j in collect(0:lim)])
```
Comment (ThijsvdLaar) [inline src/factor_nodes/poisson.jl]: Please move the definitions for the `Elog_fact()` and `h()` functions to the outer scope; there is no need to define it within the `averageEnergy()` function. Also, a more descriptive naming and specific argument typing would be appreciated.

File: src/factor_nodes/poisson.jl [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+ForneyLab.isProper(dist::ProbabilityDistribution{Univariate, Poisson}) = (0 < dist.params[:l] < huge)
+
+ForneyLab.unsafeMean(dist::ProbabilityDistribution{Univariate, Poisson}) = dist.params[:l]
+
+ForneyLab.unsafeVar(dist::ProbabilityDistribution{Univariate, Poisson}) = dist.params[:l]
+
+ForneyLab.unsafeLogMean(dist::ProbabilityDistribution{Univariate, Poisson}) = dist.params[:l]
+
+# TODO: skip the implementation of entropy and average energy for now, they require additional work
```
Comment (ThijsvdLaar) [inline src/factor_nodes/poisson.jl]: "TODO" comment can be removed

File: src/factor_nodes/poisson.jl [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+# Entropy functional
+function differentialEntropy(dist::ProbabilityDistribution{Univariate, Poisson})
+    l = clamp(dist.params[:l], tiny, huge)
```
Comment (ThijsvdLaar) [inline src/factor_nodes/poisson.jl]: Would be good to provide a reference for the results of the entropy and avg. energy in a comment together with the function definitions

File: test/factor_nodes/test_poisson.jl [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+    @test isApplicable(VBPoissonL, [ProbabilityDistribution, Nothing])
+
+    @test ruleVBPoissonL(ProbabilityDistribution(Poisson, l=2.0), nothing) == Message(Gamma, a=3.0, b=1.0)
+    @test ruleVBPoissonL(ProbabilityDistribution(Univariate, PointMass, m=2.0), nothing) == Message(Gamma, a=3.0, b=1.0)
+end
+
+@testset "averageEnergy and differentialEntropy" begin
+    # TODO: Eventually these need to be implemented, but they require additional work for the Poisson distribution
+    # @test isapprox(differentialEntropy(ProbabilityDistribution(Poisson, l=2.0)), averageEnergy(Poisson, ProbabilityDistribution(Poisson, l=2.0), ProbabilityDistribution(Univariate, PointMass, m=2.0)))
+    @test true == false
```
Comment (ThijsvdLaar) [inline test/factor_nodes/test_poisson.jl]: Please implement proper tests for the `averageEnergy` and `differentialEntropy` functions

File: src/engines/julia/update_rules/poisson.jl [hunk 6]
```diff
@@ -0,0 +1,17 @@
+export
+ruleSPPoissonOutVP,
+ruleSPPoissonLPV,
+ruleVBPoissonOut,
+ruleVBPoissonL
+
+ruleSPPoissonOutVP(msg_out::Nothing, msg_l::Message{PointMass, Univariate}) =
+    Message(Univariate, Poisson, l=deepcopy(msg_l.dist.params[:m]))
```
Comment (ThijsvdLaar) [inline src/engines/julia/update_rules/poisson.jl]: Deep copies are not needed for scalars

File: src/factor_nodes/poisson.jl [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    h(l, j=20) = sum([(l)^(k)*log(factorial(k))/factorial(k) for k in collect(0:j)])
+    l*(1-log(l)) + exp(-l)*h(l)
+end
+
+# Average energy functional
+function ForneyLab.averageEnergy(::Type{Poisson}, marg_out::ProbabilityDistribution{Univariate}, marg_l::ProbabilityDistribution{Univariate})
+    # approximation of expectation of logX!
+    h(j) = sum([binomial(j, k)*(-1)^(-k)*log(factorial(k)) for k in collect(0:j)])
```
Comment (ThijsvdLaar) [inline src/factor_nodes/poisson.jl]: The `lfact()` function implements a numerically more stable computation of the log-factorial

Comment (albertpod) [reply src/factor_nodes/poisson.jl]: Is` lfact()` a function of any specific package?

**Pull request merged**